sync/http/component.go

```diff
@@ -121,3 +122,153 @@ func (c *Component) createHTTPServer() *http.Server {
 		Handler:      routerAfterMiddleware,
 	}
 }
+
+const fieldSetMsg = "Setting property '%v' for '%v'"
+
+// Builder gathers all required and optional properties, in order
+// to construct an HTTP component.
+type Builder struct {
+	ac               AliveCheckFunc
+	rc               ReadyCheckFunc
+	httpPort         int
+	httpReadTimeout  time.Duration
+	httpWriteTimeout time.Duration
+	routes           []Route
+	middlewares      []MiddlewareFunc
+	certFile         string
+	keyFile          string
+	errors           []error
+}
+
+// NewBuilder initiates the HTTP component builder chain.
+// The builder instantiates the component using default values for
+// HTTP Port, Alive/Ready check functions and Read/Write timeouts.
+func NewBuilder() *Builder {
+	var errs []error
+	return &Builder{
+		ac:               DefaultAliveCheck,
+		rc:               DefaultReadyCheck,
+		httpPort:         httpPort,
+		httpReadTimeout:  httpReadTimeout,
+		httpWriteTimeout: httpWriteTimeout,
+		errors:           errs,
+	}
+}
+
+// WithSSL sets the filenames for the Certificate and Keyfile, in order to enable SSL.
+func (cb *Builder) WithSSL(c, k string) *Builder {
+	if c == "" || k == "" {
+		cb.errors = append(cb.errors, errors.New("Invalid cert or key provided"))
+	} else {
+		log.Info(fieldSetMsg, "Cert, Key", c+","+k)
+		cb.certFile = c
+		cb.keyFile = k
+	}
+
+	return cb
+}
+
+// WithRoutes adds routes to the HTTP component.
+func (cb *Builder) WithRoutes(rr []Route) *Builder {
```

Comment:
would it make sense to make this a varargs ? like the below method for the middlewares ?

Reply:
Yes, it could certainly be done! I'm not sure if it's just a cosmetic difference, or there's some logic underneath, but I initially just wanted compatibility with the [current OptionFunc approach](https://github.com/beatlabs/patron/blob/9b7a1fcd0a993c448f6205231e83b120de3c23a7/sync/http/option.go#L33).

What I'm seeing right now is some Patron projects defining their routes as `[]phttp.Route{}` in a centralized way as in [here](https://github.com/taxibeat/segmento/blob/b35cbf172bd45f0f722cf91f4d806a0726a47c1a/internal/infra/http/http.go#L49) and [here](https://github.com/taxibeat/rems/blob/e2c6be792086a6feb215b68a2c7eb5f3438e8a5c/internal/infra/http/http.go#L120) and passing their middlewares explicitly as in [here](https://github.com/taxibeat/rems/blob/e2c6be792086a6feb215b68a2c7eb5f3438e8a5c/cmd/rems/main.go#L220) and [here](https://github.com/taxibeat/segmento/blob/b35cbf172bd45f0f722cf91f4d806a0726a47c1a/cmd/segmento/main.go#L64).  I don't have a strong preference for either option, but somewhat like the current approach. What do you think?

Reply:
variadic arguments allow you to ommit the argument altogether which make the API unclear...

Reply:
I think it would be good to have a unified pattern for methods like these (e.g. withRoutes , withMiddlewares etc ... ). So while we are "touching" this code, it would make sense to me to apply this now. Regarding the question of slices vs. varargs, i tend to agree with @mantzas .